### PR TITLE
 fabric scoped event support

### PR DESCRIPTION
**Pull request by yunhanw-google**
#### Problem
ACLr has fabric scoped event, we need to add fabric scoped event support to handle this

#### Change overview
-- LogEvent would check if clusterObject for event has fabric, if it has, then encode it using profile tag in TLV circular buffer.
-- FetchEvent would take the new fabricIndex parameter from read handler, and then when filtering the interested event from circular event buffer, it would compare the current fabric index with the one stored for each event if it has fabric index

#### Testing
Update TestEvent cluster with fabric index and test this feature.

### CODE REVIEW

File: src/app/EventManagement.cpp [hunk 1]
```diff
@@ -747,6 +763,13 @@ CHIP_ERROR EventManagement::FetchEventParameters(const TLVReader & aReader, size
         envelope->mCurrentTime.mType  = Timestamp::Type::kEpoch;
         envelope->mCurrentTime.mValue = epochTime;
     }
+
+    if (TLV::IsProfileTag(reader.GetTag()) && reader.GetTag() == TLV::ProfileTag(kEventManagementProfile, kFabricIndexTag))
+    {
+        FabricIndex fabricIndex = 0;
```
Comment (erjiaqing) [inline src/app/EventManagement.cpp:769]: Please use kUndefinedFabricIndex instead of 0

File: src/app/EventManagement.cpp [hunk 2]
```diff
@@ -580,6 +585,9 @@ CHIP_ERROR EventManagement::CopyEvent(const TLVReader & aReader, TLVWriter & aWr
 
 static bool IsInterestedEventPaths(EventLoadOutContext * eventLoadOutContext, const EventEnvelopeContext & event)
 {
+    ChipLogError(EventLogging, "current fabric index is %u", eventLoadOutContext->mFabricIndex);
+    ;
```
Comment (erjiaqing) [inline src/app/EventManagement.cpp:589]: Nit: extra `;`

File: src/app/EventManagement.cpp [hunk 3]
```diff
@@ -590,7 +598,14 @@ static bool IsInterestedEventPaths(EventLoadOutContext * eventLoadOutContext, co
     {
         if (interestedPath->IsEventPathSupersetOf(path))
         {
-            return true;
+            if (event.mFabricIndex.IsNull())
```
Comment (erjiaqing) [inline src/app/EventManagement.cpp:601]: Seems this logic should is not related to the path? Can we move this out side the loop?

File: src/app/EventManagement.cpp [hunk 4]
```diff
@@ -580,6 +585,9 @@ CHIP_ERROR EventManagement::CopyEvent(const TLVReader & aReader, TLVWriter & aWr
 
 static bool IsInterestedEventPaths(EventLoadOutContext * eventLoadOutContext, const EventEnvelopeContext & event)
 {
+    ChipLogError(EventLogging, "current fabric index is %u", eventLoadOutContext->mFabricIndex);
```
Comment (erjiaqing) [inline src/app/EventManagement.cpp:588]: Please remove this debug message or change its priority.

File: src/app/EventLogging.h [hunk 5]
```diff
@@ -61,7 +63,46 @@ class EventLogger : public EventLoggingDelegate
...
+                    EventOptions::Type aUrgent = EventOptions::Type::kNotUrgent)
+{
+    EventLogger<T> eventData(aEventData);
+    ConcreteEventPath path(aEndpoint, aEventData.GetClusterId(), aEventData.GetEventId());
+    EventManagement & logMgmt = chip::app::EventManagement::GetInstance();
+    EventOptions eventOptions;
+    eventOptions.mUrgent      = aUrgent;
+    eventOptions.mPath        = path;
+    eventOptions.mPriority    = aEventData.GetPriorityLevel();
+    eventOptions.mFabricIndex = aEventData.GetNullableFabricIndex();
```
Comment (erjiaqing) [inline src/app/EventLogging.h:101]: Nit: Should be GetFabricIndex()

File: src/app/EventLogging.h [hunk 6]
```diff
@@ -61,7 +63,46 @@ class EventLogger : public EventLoggingDelegate
...
+{
+    EventLogger<T> eventData(aEventData);
+    ConcreteEventPath path(aEndpoint, aEventData.GetClusterId(), aEventData.GetEventId());
+    EventManagement & logMgmt = chip::app::EventManagement::GetInstance();
+    EventOptions eventOptions;
+    eventOptions.mUrgent   = aUrgent;
+    eventOptions.mPath     = path;
+    eventOptions.mPriority = aEventData.GetPriorityLevel();
+    eventOptions.mFabricIndex.SetNonNull(aEventData.GetFabricIndex());
+    ChipLogError(EventLogging, "current fabric index is %u in LogEvent", aEventData.GetFabricIndex());
```
Comment (erjiaqing) [inline src/app/EventLogging.h:81]: Please remove this log

File: src/app/EventManagement.cpp (hunk 2 above)
Comment (yunhanw-google) [reply src/app/EventManagement.cpp:589]: this is one debug log, would remove it soon.

File: src/app/EventManagement.cpp (hunk 1 above)
Comment (yunhanw-google) [reply src/app/EventManagement.cpp:771]: Done, thanks, It is ready for formal review.

File: src/app/EventManagement.cpp (hunk 2 above)
Comment (yunhanw-google) [reply src/app/EventManagement.cpp]: Done, thanks, It is ready for formal review.

File: src/app/EventManagement.cpp (hunk 3 above)
Comment (yunhanw-google) [reply src/app/EventManagement.cpp:600]: Done, thanks, It is ready for formal review.

File: src/app/EventManagement.cpp (hunk 4 above)
Comment (yunhanw-google) [reply src/app/EventManagement.cpp]: Done, thanks, It is ready for formal review.

File: src/app/EventLogging.h (hunk 6 above)
Comment (yunhanw-google) [reply src/app/EventLogging.h]: Done, thanks, It is ready for formal review.

File: src/app/EventLogging.h (hunk 5 above)
Comment (yunhanw-google) [reply src/app/EventLogging.h:99]: Done, thanks, It is ready for formal review.

File: src/app/EventManagement.cpp [hunk 7]
```diff
@@ -406,25 +411,27 @@ CircularEventBuffer * EventManagement::GetPriorityBuffer(PriorityLevel aPriority
...
+                                  ctx->mpContext->mCurrentTime.mValue - ctx->mpContext->mPreviousTime.mValue);
     }
     else if ((aReader.GetTag() == TLV::ContextTag(to_underlying(EventDataIB::Tag::kEpochTimestamp))) && !(ctx->mpContext->mFirst))
     {
-        err = ctx->mpWriter->Put(TLV::ContextTag(to_underlying(EventDataIB::Tag::kDeltaEpochTimestamp)),
-                                 ctx->mpContext->mCurrentTime.mValue - ctx->mpContext->mPreviousTime.mValue);
+        return ctx->mpWriter->Put(TLV::ContextTag(to_underlying(EventDataIB::Tag::kDeltaEpochTimestamp)),
+                                  ctx->mpContext->mCurrentTime.mValue - ctx->mpContext->mPreviousTime.mValue);
+    }
+    else if (TLV::IsProfileTag(aReader.GetTag()) && aReader.GetTag() == TLV::ProfileTag(kEventManagementProfile, kFabricIndexTag))
```
Comment (erjiaqing) [inline src/app/EventManagement.cpp:427]: Is it possible to remove this branch by adding a condition below so we can know what is expected to be copied?

Comment (erjiaqing) [reply src/app/EventManagement.cpp:427]: Or can we move this to the very beginning of the check so we can know that the fabric index is an internal field and will never be copied?

File: src/app/EventLogging.h [hunk 8]
```diff
@@ -53,15 +55,55 @@ class EventLogger : public EventLoggingDelegate
  * context tags to be interpreted within the schema identified by
  * `ClusterID` and `EventId`.
  *
- * @param[in] apDelegate The EventLoggingDelegate to serialize the event data
+ * LogEvent has 3 variant, for aEventData that have method GetFabricIndex with return type FabricIndex or
+ * DataModel::Nullable<chip::FabricIndex>, it would use first two logEvent function twhich can process the fabric index, for
```
Comment (yufengwangca) [inline src/app/EventLogging.h:59]: typo: logEvent function which can process

File: src/app/EventLogging.h [hunk 9]
```diff
@@ -53,15 +55,55 @@ class EventLogger : public EventLoggingDelegate
...
+ * aEventData that don't have GetFabricIndex, it would use tird LogEvent function.
```
Comment (yufengwangca) [inline src/app/EventLogging.h:60]: typo: tird -> third

Comment (bzbarsky-apple) [inline src/app/EventLogging.h:60]: When would a null FabricIndex be returned?  How is that different from a FabricIndex of 0?

(Also, is there a reason this is using Nullable instead of Optional to mean "no fabric index"?  Nullable is meant to be used for the spec concept of nullable; that being things that get encoded as a TLV null.  Do we do that for event fabric indices?)

File: src/app/EventLogging.h [hunk 10]
```diff
@@ -53,15 +55,55 @@ class EventLogger : public EventLoggingDelegate
...
  * @param[in] aUrgent    The EventOption Type, kUrgent or kNotUrgent
  * @param[out] aEventNumber The event Number if the event was written to the
  *                         log, 0 otherwise. The Event number is expected to monotonically increase.
  *
  * @return CHIP_ERROR  CHIP Error Code
  */
-template <typename T>
+template <typename T,
+          typename std::enable_if_t<std::is_class<T>::value && DataModel::IsFabricScoped<T>::value &&
+                                        std::is_same<decltype(std::declval<T>().GetFabricIndex()), FabricIndex>::value,
```
Comment (bzbarsky-apple) [inline src/app/EventLogging.h:71]: What other sort of type would GetFabricIndex be returning?

File: src/app/EventLoggingTypes.h [hunk 11]
```diff
@@ -140,6 +141,7 @@ class EventOptions
     PriorityLevel mPriority = PriorityLevel::Invalid;
     Type mUrgent            = Type::kNotUrgent; /**< A flag denoting if the event is time sensitive.  When kUrgent is set, it causes
                                                        the event log to be flushed. */
+    DataModel::Nullable<chip::FabricIndex> mFabricIndex;
```
Comment (bzbarsky-apple) [inline src/app/EventLoggingTypes.h:144]: This needs to very clearly document what the semantics of mFabricIndex are (esp. if it's null).

File: src/app/EventManagement.cpp [hunk 12]
```diff
@@ -97,6 +97,7 @@ struct EventEnvelopeContext
     EndpointId mEndpointId   = 0;
     EventId mEventId         = 0;
     EventNumber mEventNumber = 0;
+    DataModel::Nullable<chip::FabricIndex> mFabricIndex;
```
Comment (bzbarsky-apple) [inline src/app/EventManagement.cpp:100]: Again, what does it mean for this to be null?

File: src/app/EventManagement.cpp [hunk 13]
```diff
@@ -359,6 +360,10 @@ CHIP_ERROR EventManagement::ConstructEvent(EventLoadOutContext * apContext, Even
 
     // Callback to write the EventData
     ReturnErrorOnFailure(apDelegate->WriteEvent(apContext->mWriter));
+    if (!(apOptions->mFabricIndex.IsNull()))
```
Comment (bzbarsky-apple) [inline src/app/EventManagement.cpp:363]: OK, so this is definitely an `Optional`, not a `Nullable`....

File: src/app/EventManagement.cpp [hunk 14]
```diff
@@ -359,6 +360,10 @@ CHIP_ERROR EventManagement::ConstructEvent(EventLoadOutContext * apContext, Even
...
+    {
+        apContext->mWriter.Put(TLV::ProfileTag(kEventManagementProfile, kFabricIndexTag), apOptions->mFabricIndex.Value());
```
Comment (bzbarsky-apple) [inline src/app/EventManagement.cpp:365]: Worth documenting that this TLV is not going on the wire directly, so the fact that we're making up some random profile tag is ok because we are the only ones who will see it.

File: src/app/EventManagement.cpp (hunk 7 above)
Comment (bzbarsky-apple) [inline src/app/EventManagement.cpp:427]: That IsProfileTag() seems redundant with the equality check....

File: src/app/EventManagement.cpp [hunk 15]
```diff
@@ -584,6 +592,12 @@ static bool IsInterestedEventPaths(EventLoadOutContext * eventLoadOutContext, co
     {
         return false;
     }
+
+    if (!event.mFabricIndex.IsNull() && eventLoadOutContext->mFabricIndex != event.mFabricIndex.Value())
```
Comment (bzbarsky-apple) [inline src/app/EventManagement.cpp:596]: OK, so here is presumably where we don't want to just use a 0 fabric index to mean "not fabric associated at all"?  Or do we not?  For an event that may or may not be fabric-scoped, how does one indicate "not associated with a particular fabric" in the spec?

File: src/app/EventManagement.cpp [hunk 16]
```diff
@@ -661,7 +675,7 @@ CHIP_ERROR EventManagement::CopyEventsSince(const TLVReader & aReader, size_t aD
 }
 
 CHIP_ERROR EventManagement::FetchEventsSince(TLVWriter & aWriter, ClusterInfo * apClusterInfolist, EventNumber & aEventMin,
-                                             size_t & aEventCount)
+                                             size_t & aEventCount, const FabricIndex & aFabricIndex)
```
Comment (bzbarsky-apple) [inline src/app/EventManagement.cpp:678]: Why pass FabricIndex by reference?

File: src/app/EventManagement.cpp [hunk 17]
```diff
@@ -747,6 +762,13 @@ CHIP_ERROR EventManagement::FetchEventParameters(const TLVReader & aReader, size
         envelope->mCurrentTime.mType  = Timestamp::Type::kEpoch;
         envelope->mCurrentTime.mValue = epochTime;
     }
+
+    if (TLV::IsProfileTag(reader.GetTag()) && reader.GetTag() == TLV::ProfileTag(kEventManagementProfile, kFabricIndexTag))
```
Comment (bzbarsky-apple) [inline src/app/EventManagement.cpp:766]: Again, the IsProfileTag() is redundant.

File: src/app/data-model/FabricScopedEvent.h [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+
+#include <lib/support/TypeTraits.h>
+#include <type_traits>
+
+namespace chip {
+namespace app {
+namespace DataModel {
+
+/*
+ * Check whether a cluster object event is fabric scoped.
```
Comment (bzbarsky-apple) [inline src/app/data-model/FabricScopedEvent.h:29]: Please coordinate with #14227.

File: src/app/zap-templates/zcl/data-model/chip/test-cluster.xml [hunk 19]
```diff
@@ -520,5 +550,16 @@ limitations under the License.
...
     </event>
+
+    <event code="0x0002" name="TestFabricScopedEvent" priority="info" side="server">
+      <description>Example test event</description>
+      <field id="1" name="arg1" type="fabric_idx"/>
+    </event>
+
+    <event code="0x0003" name="TestNullableFabricScopedEvent" priority="info" side="server">
+      <description>Example test event</description>
+      <field id="1" name="arg1" type="fabric_idx" isNullable="true"/>
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/chip/test-cluster.xml:561]: What spec construct does this correspond to.  Why are we caring about this case?

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src/app/zap-templates/zcl/data-model/chip/test-cluster.xml (hunk 19 above)
Comment (yunhanw-google) [reply src/app/zap-templates/zcl/data-model/chip/test-cluster.xml:561]: For events in ACL cluster, it has FabricIndex, we generate code FabricIndex GetFabricIndex() const { return adminFabricIndex; } where adminFabricIndex is not nullable type, its type is FabricIndex.
For events in doorlock cluster, it has nullable fabricindex, we generate DataModel::Nullable<chip::FabricIndex>  etFabricIndex() const { return fabricIndex; }  where fabricindex is not nullable type. 
Then for LogEvent, I wrote two kind of template function to correspond to these two situation, so in order to validate these two situation, I have type="fabric_idx" isNullable="true".

This is nullable fabric index from doorlock cluster
https://github.com/project-chip/connectedhomeip/blob/master/src/app/zap-templates/zcl/data-model/chip/door-lock-cluster.xml#L627

File: src/app/EventManagement.cpp (hunk 7 above)
Comment (yunhanw-google) [reply src/app/EventManagement.cpp:427]: Sure, just move it to the beginning, thanks

File: src/app/EventLogging.h (hunk 8 above)
Comment (yunhanw-google) [reply src/app/EventLogging.h:59]: Done, thanks

File: src/app/EventLogging.h (hunk 9 above)
Comment (yunhanw-google) [reply src/app/EventLogging.h:60]: Done, thanks

File: src/app/EventLogging.h (hunk 10 above)
Comment (yunhanw-google) [reply src/app/EventLogging.h:71]: Two king of returned.
In acl cluster, for event with admin fabric index without nullable, it would have
FabricIndex GetFabricIndex() const { return adminFabricIndex; } where adminFabricIndex is not nullable type, its type is FabricIndex

In doorlock cluster, for event with fabric index with nullable, it would have
DataModel::Nullablechip::FabricIndex etFabricIndex() const { return fabricIndex; } where fabricindex is not nullable type.

File: src/app/EventLoggingTypes.h (hunk 11 above)
Comment (yunhanw-google) [reply src/app/EventLoggingTypes.h:144]: from spec, "The fabric index of the fabric that performed the lock operation. This SHALL be null if there is no fabric that can be determined for the given operation source. This SHALL NOT be null if the operation source is "Remote".,
For null, I think it means this event can be retrieved by read handler, and there is no fabric restriction on it.

File: src/app/EventLogging.h (hunk 9 above)
Comment (yunhanw-google) [reply src/app/EventLogging.h:60]: from spec, https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/686ca55b4764cc6e1fccd348671a7360a5f6828e/src/app_clusters/DoorLock.adoc#534-fabricindex."The fabric index of the fabric that performed the lock operation. This SHALL be null if there is no fabric that can be determined for the given operation source. This SHALL NOT be null if the operation source is "Remote".,
For null, I think it means this event can be retrieved by read handler, and there is no fabric restriction on it. 0 is still meaningful fabric.

Comment (yunhanw-google) [reply src/app/EventLogging.h:60]: Updated document.

File: src/app/EventManagement.cpp (hunk 12 above)
Comment (yunhanw-google) [reply src/app/EventManagement.cpp:100]: Updated doc in EventLogging.h.
For nullable fabric index, it means this event can be retrieved by read handler without the fabric restriction. 0 is still meaningful fabric.

File: src/app/EventManagement.cpp (hunk 13 above)
Comment (yunhanw-google) [reply src/app/EventManagement.cpp:363]: Currently these LogEvent template function are unifying using nullable fabricIndex, preferring still to use nullable fabricIndex here, thanks

File: src/app/EventManagement.cpp (hunk 14 above)
Comment (yunhanw-google) [reply src/app/EventManagement.cpp:365]: Update document, thanks

File: src/app/EventManagement.cpp (hunk 15 above)
Comment (yunhanw-google) [reply src/app/EventManagement.cpp:596]: Yes, since we have nullable fabric index, and in door lock cluster spec, it says if fabric index is null, the generated code is using DataModel::Nullable<FabricIndex>, I prefer use nullable to represent "not fabric associated at all". https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/686ca55b4764cc6e1fccd348671a7360a5f6828e/src/app_clusters/DoorLock.adoc#534-fabricindex

File: src/app/EventManagement.cpp (hunk 17 above)
Comment (yunhanw-google) [reply src/app/EventManagement.cpp]: removed, thanks

File: src/app/EventManagement.cpp (hunk 16 above)
Comment (yunhanw-google) [reply src/app/EventManagement.cpp:681]: I assume every handler would have one fabric index associated, I expect FetchEventsSince would always take one meaningful fabric index. thanks

File: src/app/zap-templates/zcl/data-model/chip/test-cluster.xml (hunk 19 above)
Comment (bzbarsky-apple) [reply src/app/zap-templates/zcl/data-model/chip/test-cluster.xml:561]: > For events in doorlock cluster, it has nullable fabricindex

Those are NOT FABRIC SCOPED events.  They are just events that happen to have a fabricIndex that does NOT scope them, and they must be readable by all fabrics.  Which fabric-scoped events are not.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]
Sounds like this completely breaks door lock physical security by not allowing a fabric other than the one that unlocks the door to know that the door was unlocked.

File: src/app/EventLoggingTypes.h (hunk 11 above)
Comment (bzbarsky-apple) [reply src/app/EventLoggingTypes.h:144]: All door lock events have no fabric restrictions.

File: src/app/EventManagement.cpp (hunk 12 above)
Comment (bzbarsky-apple) [reply src/app/EventManagement.cpp:100]: How is 0 a "meaningful fabric"?

File: src/app/EventLogging.h (hunk 9 above)
Comment (yunhanw-google) [reply src/app/EventLogging.h]: Per offline chat, we would check for a field that is of type fabric_idx and not nullable which could be used for fabric-scoped event, currently acl is fabric scope one. updated the code

File: src/app/EventLoggingTypes.h (hunk 11 above)
Comment (yunhanw-google) [reply src/app/EventLoggingTypes.h:144]: Per offline chat, we would check for a field that is of type fabric_idx and not nullable which could be used for fabric-scoped event, currently acl is fabric scope one. updated the code

File: src/app/EventManagement.cpp (hunk 13 above)
Comment (yunhanw-google) [reply src/app/EventManagement.cpp:366]: update with nullable

File: src/app/EventManagement.cpp (hunk 12 above)
Comment (yunhanw-google) [reply src/app/EventManagement.cpp:100]: updated. use 0 as undefined spec. remove nullable

File: src/app/EventManagement.cpp (hunk 15 above)
Comment (yunhanw-google) [reply src/app/EventManagement.cpp:599]: Per offline chat, we would check for a field that is of type fabric_idx and not nullable which could be used for fabric-scoped event, currently acl is fabric scope one. updated the code

File: src/app/zap-templates/zcl/data-model/chip/test-cluster.xml (hunk 19 above)
Comment (yunhanw-google) [reply src/app/zap-templates/zcl/data-model/chip/test-cluster.xml:561]: Per offline chat, we would check for a field that is of type fabric_idx and not nullable which could be used for fabric-scoped event, currently acl is fabric scope one. updated the code

File: src/app/EventManagement.cpp (hunk 7 above)
Comment (yunhanw-google) [reply src/app/EventManagement.cpp]: removed. thanks

**Comment by yunhanw-google** [comment]
> Sounds like this completely breaks door lock physical security by not allowing a fabric other than the one that unlocks the door to know that the door was unlocked.

Per offline chat, we would check for a field that is of type fabric_idx and not nullable which could be used for fabric-scoped event, currently only acl is fabric scope one. updated the code

File: examples/lighting-app/lighting-common/lighting-app.matter [hunk 20]
```diff
@@ -1012,6 +1012,21 @@ client cluster OnOff = 6 {
   attribute enum8 startUpOnOff = 16387;
   readonly global attribute bitmap32 featureMap = 65532;
   readonly global attribute int16u clusterRevision = 65533;
+
+  request struct OffWithEffectRequest {
```
Comment (bzbarsky-apple) [inline examples/lighting-app/lighting-common/lighting-app.matter:1016]: Should probably not have changes in here.

File: src/app/EventLogging.h [hunk 21]
```diff
@@ -22,7 +22,9 @@
 #include <app/EventLoggingDelegate.h>
 #include <app/EventManagement.h>
 #include <app/data-model/Encode.h>
+#include <app/data-model/FabricScoped.h>
 #include <app/data-model/List.h> // So we can encode lists
+#include <app/data-model/Nullable.h>
```
Comment (bzbarsky-apple) [inline src/app/EventLogging.h:27]: Not used; please don't add.

File: src/app/EventLogging.h [hunk 22]
```diff
@@ -53,15 +55,36 @@ class EventLogger : public EventLoggingDelegate
  * context tags to be interpreted within the schema identified by
  * `ClusterID` and `EventId`.
  *
- * @param[in] apDelegate The EventLoggingDelegate to serialize the event data
+ * LogEvent has 2 variant, for aEventData that have method GetFabricIndex with return type FabricIndex,
+ * it would use first  logEvent function which can process the fabric index, for
+ * aEventData that don't have GetFabricIndex, it would use second LogEvent function.
```
Comment (bzbarsky-apple) [inline src/app/EventLogging.h:60]:
```suggestion
 * LogEvent has 2 variant, one for fabric-scoped events and one for non-fabric-scoped events.
```

File: src/app/EventLogging.h [hunk 23]
```diff
@@ -53,15 +55,36 @@ class EventLogger : public EventLoggingDelegate
...
+                    EventOptions::Type aUrgent = EventOptions::Type::kNotUrgent)
+{
+    EventLogger<T> eventData(aEventData);
+    ConcreteEventPath path(aEndpoint, aEventData.GetClusterId(), aEventData.GetEventId());
+    EventManagement & logMgmt = chip::app::EventManagement::GetInstance();
+    EventOptions eventOptions;
+    eventOptions.mUrgent   = aUrgent;
+    eventOptions.mPath     = path;
+    eventOptions.mPriority = aEventData.GetPriorityLevel();
+    if (aEventData.GetFabricIndex() != kUndefinedFabricIndex)
```
Comment (bzbarsky-apple) [inline src/app/EventLogging.h:80]: Why?  What cases is this trying to handle, and why is handling them by treating those events as not fabric scoped the right thing?

File: src/app/EventLoggingTypes.h [hunk 24]
```diff
@@ -18,6 +18,7 @@
 #pragma once
 
 #include <app/ClusterInfo.h>
+#include <app/data-model/Nullable.h>
```
Comment (bzbarsky-apple) [inline src/app/EventLoggingTypes.h:21]: Should be including Optional.h instead.

File: src/app/EventLoggingTypes.h [hunk 25]
```diff
@@ -140,6 +141,7 @@ class EventOptions
     PriorityLevel mPriority = PriorityLevel::Invalid;
     Type mUrgent            = Type::kNotUrgent; /**< A flag denoting if the event is time sensitive.  When kUrgent is set, it causes
                                                        the event log to be flushed. */
+    Optional<FabricIndex> mFabricIndex;
```
Comment (bzbarsky-apple) [inline src/app/EventLoggingTypes.h:144]: Should document the semantics of this member, especially what it means to have or not have a value.

File: src/app/EventManagement.cpp [hunk 26]
```diff
@@ -97,6 +97,7 @@ struct EventEnvelopeContext
     EndpointId mEndpointId   = 0;
     EventId mEventId         = 0;
     EventNumber mEventNumber = 0;
+    Optional<FabricIndex> mFabricIndex;
```
Comment (bzbarsky-apple) [inline src/app/EventManagement.cpp:100]: Again, document.  Maybe by just pointing to the documentation for EventOptions?

File: src/app/EventManagement.cpp [hunk 27]
```diff
@@ -406,25 +414,27 @@ CircularEventBuffer * EventManagement::GetPriorityBuffer(PriorityLevel aPriority
...
 CHIP_ERROR EventManagement::CopyAndAdjustDeltaTime(const TLVReader & aReader, size_t aDepth, void * apContext)
 {
-    CHIP_ERROR err                      = CHIP_NO_ERROR;
     CopyAndAdjustDeltaTimeContext * ctx = static_cast<CopyAndAdjustDeltaTimeContext *>(apContext);
     TLVReader reader(aReader);
 
-    if ((aReader.GetTag() == TLV::ContextTag(to_underlying(EventDataIB::Tag::kSystemTimestamp))) && !(ctx->mpContext->mFirst))
+    if (aReader.GetTag() == TLV::ProfileTag(kEventManagementProfile, kFabricIndexTag))
+    {
+        return CHIP_NO_ERROR;
```
Comment (bzbarsky-apple) [inline src/app/EventManagement.cpp:422]:
```suggestion
        // Does not go on the wire.
        return CHIP_NO_ERROR;
```

File: src/app/EventManagement.cpp [hunk 28]
```diff
@@ -584,6 +595,12 @@ static bool IsInterestedEventPaths(EventLoadOutContext * eventLoadOutContext, co
     {
         return false;
     }
+
+    if (event.mFabricIndex.HasValue() && eventLoadOutContext->mFabricIndex != event.mFabricIndex.Value())
```
Comment (bzbarsky-apple) [inline src/app/EventManagement.cpp:599]: So in particular, the way things are right now any events generated by a PASE session before it's picked up a fabric association will be visible to all consumers.

Maybe that's not a problem in practice because you can't add ACLS via a PASE session before it picks up a fabric?  But we would need to pretty carefully check the ordering of things in AddNOC....

On the other hand, if we allowed event.mFabricIndex to have the "no fabric" value (0).... I guess a later PASE session would be able to see the events, which is not desirable?

Maybe the right answer is that fabric-scoped events emitted not in the context of a fabric should just be dropped completely?  Worth checking with @mlepage-google and @tcarmelveilleux whether they have thoughts on this.  But we can also adjust it as a followup.

File: src/app/EventManagement.cpp [hunk 29]
```diff
@@ -661,7 +678,7 @@ CHIP_ERROR EventManagement::CopyEventsSince(const TLVReader & aReader, size_t aD
 }
 
 CHIP_ERROR EventManagement::FetchEventsSince(TLVWriter & aWriter, ClusterInfo * apClusterInfolist, EventNumber & aEventMin,
-                                             size_t & aEventCount)
+                                             size_t & aEventCount, const FabricIndex & aFabricIndex)
```
Comment (bzbarsky-apple) [inline src/app/EventManagement.cpp:681]: No, my question was why the last arg here is not `FabricIndex aFabricIndex`.  As in, why are we passing a reference, not a value.

File: src/app/zap-templates/templates/app/cluster-objects.zapt [hunk 30]
```diff
@@ -206,6 +206,14 @@ public:
     {{zapTypeToEncodableClusterObjectType type}} {{asLowerCamelCase name}}{{> cluster_objects_field_init}};
     {{/zcl_event_fields}}
 
+    {{#if event_is_fabric_scoped}}
+    {{#unless event_is_fabric_index_nullable}}
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/templates/app/cluster-objects.zapt:210]: This does not look right.  `event_is_fabric_scoped` should be false for events that are not fabric-scoped!

Please do a followup to fix ZAP here and then fix this logic.

**Review by bzbarsky-apple — APPROVED** [review]

File: src/app/EventManagement.cpp (hunk 28 above)
Comment (mlepage-google) [reply src/app/EventManagement.cpp:599]: Thanks I took a note to think about this.

File: src/app/EventManagement.cpp (hunk 29 above)
Comment (yunhanw-google) [reply src/app/EventManagement.cpp:681]: I think it is good to pass reference, update it. thanks

File: src/app/EventLogging.h (hunk 21 above)
Comment (yunhanw-google) [reply src/app/EventLogging.h:27]: yep, removed, thanks

File: src/app/zap-templates/templates/app/cluster-objects.zapt (hunk 30 above)
Comment (yunhanw-google) [reply src/app/zap-templates/templates/app/cluster-objects.zapt:210]: yep, would follow-up it after fixing zap, thanks

Comment (yunhanw-google) [reply src/app/zap-templates/templates/app/cluster-objects.zapt:210]: https://github.com/project-chip/connectedhomeip/issues/14365

File: src/app/EventManagement.cpp (hunk 28 above)
Comment (yunhanw-google) [reply src/app/EventManagement.cpp:599]: https://github.com/project-chip/connectedhomeip/issues/14366

File: src/app/EventManagement.cpp (hunk 27 above)
Comment (yunhanw-google) [reply src/app/EventManagement.cpp:422]: Done, thanks

File: src/app/EventLoggingTypes.h (hunk 24 above)
Comment (yunhanw-google) [reply src/app/EventLoggingTypes.h:21]: Yep, fixed, thanks

File: src/app/EventLogging.h (hunk 22 above)
Comment (yunhanw-google) [reply src/app/EventLogging.h:60]: Done, thanks

File: examples/lighting-app/lighting-common/lighting-app.matter (hunk 20 above)
Comment (yunhanw-google) [reply examples/lighting-app/lighting-common/lighting-app.matter:1016]: yep, accidental zap generated...

File: src/app/EventLogging.h (hunk 23 above)
Comment (yunhanw-google) [reply src/app/EventLogging.h:80]: right, just remove this check, and also remove optional fabricindex in this PR, I think it is fine to use 0, kUndefinedFabricIndex for "no fabric"

File: src/app/EventLoggingTypes.h (hunk 25 above)
Comment (yunhanw-google) [reply src/app/EventLoggingTypes.h:144]: just remove optional, thanks

Comment (yunhanw-google) [reply src/app/EventLoggingTypes.h:144]: documented

File: src/app/EventManagement.cpp (hunk 26 above)
Comment (yunhanw-google) [reply src/app/EventManagement.cpp:100]: done, thanks

**Review by yufengwangca — APPROVED** [review]

**Review by jmartinez-silabs — APPROVED** [review]

**Comment by mlepage-google** [comment]
I tried this with Access Control.

In one REPL, I set up two controllers on fabric 1, one subscribed to this event on fabric 1.
In another REPL, I set up a controller on fabric 2, subscribed to this event on fabric 2.

In the first REPL, I wrote the ACL attribute with a new list value.
In the first REPL, I received event notifications for changed entries.
In the second REPL, I received no such notifications.
